core/src/main/scala/org/http4s/headers/Content-Length.scala

```diff
@@ -9,21 +9,29 @@ import org.http4s.util.Writer
   *
   * The HTTP RFCs do not specify a maximum length.  We have decided that `Long.MaxValue`
   * bytes ought to be good enough for anybody in order to avoid the irritations of `BigInt`.
-  *
+  *t
   * @param length the length; throws an `IllegalArgumentException` if negative
   */
-final case class `Content-Length`(length: Long) extends Header.Parsed {
-  require(length >= 0L)
-
+sealed abstract case class `Content-Length`(length: Long) extends Header.Parsed {
   override def key: `Content-Length`.type = `Content-Length`
   override def renderValue(writer: Writer): writer.type = writer.append(length)
+  def modify(f: Long => Long): Option[`Content-Length`] = `Content-Length`.apply(f(length)).right.toOption
 }
 
 object `Content-Length` extends HeaderKey.Internal[`Content-Length`] with HeaderKey.Singleton {
+  private class ContentLengthImpl(length: Long) extends `Content-Length`(length)
+
+  val zero: `Content-Length` = new ContentLengthImpl(0) {}
+
+  def apply(length: Long): ParseResult[`Content-Length`] = fromLong(length)
```

Comment:
I wonder, is `ParseResult` really right when we're not parsing from a String?  Maybe we could call this `fromLong`, and then the old `apply` signature could remain for a deprecation cycle.

I think I could argue that `apply` should be reserved for constructors that are total.